components/script/dom/document.rs

```diff
@@ -1285,6 +1279,114 @@ impl Document {
+                        // Spin the event loop. We will be called again by
+                        // XXX when this condition changes.
+                        return;
+                    }
+
+                    {
+                        let (ref element, ref script) = deferred_scripts[0];
+                        let script = match *script {
+                            None => {
+                                // Spin the event loop. We will be called again by
+                                // ScriptContext::response_complete when this
+                                // condition changes.
+                                return;
+                            },
+                            Some(ref script) => script.clone(),
+                        };
+
+                        // Step 3.2.
+                        element.execute(script);
+                    }
+
+                    // Step 3.3.
+                    deferred_scripts.remove(0);
+                }
+            }
+
+            // Step 4.
+            self.dispatch_dom_content_loaded();
+        }
+
+        if self.end_state.get() == EndState::FiredDOMContentLoaded {
+            // Step 5.
+            if !self.asap_scripts_set.borrow().is_empty() {
+                // Spin the event loop. We will be called again by
+                // ScriptContext::response_complete when this condition changes.
+                return;
+            }
+
+            if !self.asap_in_order_scripts_list.borrow().is_empty() {
+                // Spin the event loop. We will be called again by
+                // ScriptContext::response_complete when this condition changes.
+                return;
+            }
+
+            self.end_state.set(EndState::ExecutedAsyncScripts);
+        }
+
+        if self.end_state.get() == EndState::ExecutedAsyncScripts {
+            // Step 6.
+            if self.loader().is_blocked() {
+                // Spin the event loop. We will be called again by
+                // XXX when this condition changes.
+                return;
+            }
+
+            // Step 7.
+            let window = self.window();
+            let document = Trusted::new(self, window.dom_manipulation_task_source());
+            let handler = box DocumentProgressHandler::new(document);
+            let message = MainThreadScriptMsg::Common(
+                CommonScriptMsg::RunnableMsg(ScriptThreadEventCategory::DocumentEvent, handler));
+            window.main_thread_script_chan().send(message).unwrap();
+
+            // TODO: Step 8: pageshow event.
+            // TODO: Step 9: appcache.
+            // TODO: Step 10: printing.
+            // TODO: Step 11: pageshow event.
+            // TODO: Step 12: ready for post-load tasks.
+            // TODO: Step 13: completely loaded.
+
+            let ConstellationChan(ref sender) = window.constellation_chan();
+            sender.send(ConstellationMsg::LoadComplete(window.pipeline())).unwrap();
```

Comment:
I wonder if this should be triggered as a result of running DocumentProgressHandler instead? Seems like it would be easy to cause races here.